utils/extract.go

```diff
@@ -10,65 +10,69 @@ import (
 	"io"
 	"os"
 	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
 )
 
+func sanitize(path string) (string, error) {
+	path = filepath.Clean(path)
+
+	if strings.HasPrefix(path, "..") {
+		return "", errors.New("unexpected relative path")
+	}
+
+	return path, nil
+}
+
 // ExtractTarGz takes in an io.Reader containing the bytes for a .tar.gz file and
 // a destination string to extract to.
 func ExtractTarGz(gzipStream io.Reader, dst string) error {
 	uncompressedStream, err := gzip.NewReader(gzipStream)
 	if err != nil {
-		return fmt.Errorf("ExtractTarGz: NewReader failed: %s", err.Error())
+		return errors.Wrap(err, "failed to initialize gzip reader")
 	}
 	defer uncompressedStream.Close()
 
 	tarReader := tar.NewReader(uncompressedStream)
 
 	for {
 		header, err := tarReader.Next()
-
 		if err == io.EOF {
 			break
+		} else if err != nil {
+			return errors.Wrap(err, "failed to read next file from archive")
 		}
 
+		headerName, err := sanitize(header.Name)
```

Comment:
Sanitizing before checking the type flag might not be the best option for compatibility. Archives created with non-GNU tar might contain vendor extensions and break here.